clarity-ext-scripts/clarity_ext_scripts/covid/ct_discriminator.py

```diff
@@ -0,0 +1,66 @@
+from clarity_ext.domain.validation import UsageError
+
+
+VALID_ASSAYS = ['Assay 1']
+
+BOUNDARY_DICT = {'Assay 1': (10, 50)}
+
+
+class CtDiscriminator(object):
+    """
+    A service to analyze if a sample is positive for Covid19
+    Given:
+    CT value of sample
+    Discriminator for assay (lower bound, upper bound)
+    This service does not update any artifact from context, it only reads from it
+    """
+    def __init__(self, context):
+        self.context = context
+        self.initial_validation()
+
+    def initial_validation(self):
+        if not self._has_assay_udf():
+            raise UsageError("The udf 'Assay' must be filled in before running this script")
+        if self.assay not in VALID_ASSAYS:
+            raise UsageError("The current assay value is not recognized: {}"
+                             .format(self.assay))
+
+    @property
+    def assay(self):
+        return self.context.current_step.udf_assay
+
+    def _has_assay_udf(self):
+        try:
+            dummy = self.assay
+        except AttributeError:
+            return False
+
+        return True
+
+    def _get_discriminator_values(self):
+        """
+        Returns: lower bound, upper bound
+        ct <= lower bound: counted as a positive value
+        ct >=upper bound: is counted as a negative value
+        lower bound < ct < upper bound: undefined
```

Comment:
I'm not sure that the bounds, as defined here, is actually how things will work here. Reading through the details of this, I'm not even sure that having the Ct value is enough to make the call about the diagnosis.

From the manual Maike provided:

**Quality control**
 - Blank control: Ct values at FAM and VIC/HEX channels are 0 or no data available. 
 - Positive control：Standard curves at channel FAM and VIC/HEX channels are in S-shape with Ct values not higher than 32. 
 - Testing specimen: Standard curves at VIC/HEX channel is in S-shape with Ct not higher than 32. 
 - Above requirements should be met in a single test. Otherwise, the test is invalid. Please operate the retest strictly in line with the package insert. 

**Threshold and reference**
 - rangeCut-off value of the kit was determined based on the Receiver Operator characteristic curve from testing clinical samples. Ct value for 2019-nCoV positive by the kit is not high than 38.

**Testing result interpretation**
 - The specimen is positive of 2019-nCoV if standard curve at FAM channel is in S-shape with Ct value not higher than 38. 
 - The specimen is negative of 2019-nCoV if standard curve at FAM channel is not in S-shape with Ct at FAM as 0 or no data available while Ct at VIC/HEX not higher than 32.
 - The specimen should be retested if standard curve at FAM is in S-shape with Ct higher than 38. The specimen can be reported on basis of retesting results as positive of 2019-nCoV for Ct higher than 38 and as negative of 2019-nCoV for standard curve not in S-shape and Ct of internal reference not higher than 32 at VIC/HEX. 
 - In case that standard curve at FAM is not in S-shape with Ct value as 0 or no data available, the specimen should be retested if Ct at VIC is higher than 32 or no data available.

Reply:
Maybe @kleppari wants to join this discussion?